File: tools/snippet-cutter/src/trailerFilename.ts
// Duplikat von shared/trailerFilename.ts (Repo-Root) - siehe dessen
// Kopf-Kommentar: dieses Tool hat bewusst KEINE Laufzeit-Abhaengigkeit zum
// restlichen Repo (kein npm-Workspace-Mitglied, laeuft komplett
// eigenstaendig lokal), deshalb ist das Muster hier bewusst dupliziert statt
// importiert. Bei einer Aenderung MUSS diese Kopie synchron mit
// shared/trailerFilename.ts und backend/src/services/trailerScan.ts
// gehalten werden.
//
// Erwartetes Dateinamensmuster (ohne Extension):
//   trailer-<Filmtitel> (<Jahr>) {imdb-id ttXXXXXXX}
// Rechtsverankert (das `$` am Ende), damit Klammern innerhalb des
// Filmtitels selbst (z.B. "(T)Raumschiff Surprise - Periode 1") die
// Erkennung von Jahr/imdb-id am Ende nicht stoeren.
export const TRAILER_FILENAME_PATTERN = /^trailer-(.+?)\s\((\d{4})\)\s\{imdb-id\s+(tt\d+)\}$/;

export interface ParsedTrailerFilename {
  title: string;
  year: number;
  imdbId: string;
}

export function parseTrailerFilename(basenameWithoutExt: string): ParsedTrailerFilename | null {
  const match = TRAILER_FILENAME_PATTERN.exec(basenameWithoutExt);
  if (!match) return null;
  const [, title, yearRaw, imdbId] = match;
  return { title, year: Number(yearRaw), imdbId };
}

/** Baut aus den geparsten Teilen wieder exakt denselben Dateinamens-Stamm
 * (ohne Extension) - haelt den Zieldateinamen bit-identisch zum
 * Quelldateinamen (nur die Extension wechselt zu .mp4), damit der
 * Backend-Scan (trailerScan.ts) dieselbe Regex/dasselbe Namensschema
 * nutzen kann. */
export function buildTrailerFilenameStem(parsed: ParsedTrailerFilename): string {
  return `trailer-${parsed.title} (${parsed.year}) {imdb-id ${parsed.imdbId}}`;
}

File: tools/snippet-cutter-gui/TrailerFilename.cs
using System.Text.RegularExpressions;

namespace SnippetCutterGui;

// C#-Portierung von shared/trailerFilename.ts (siehe dort fuer Details zum
// Namensmuster) - dieses GUI-Tool ist bewusst ein eigenstaendiges .NET-Projekt
// ohne Node-Laufzeitabhaengigkeit, deshalb Kopie statt Shared-Referenz. Bei
// Aenderungen am Namensmuster muss diese Datei synchron zu
// shared/trailerFilename.ts und tools/snippet-cutter/src/trailerFilename.ts
// gehalten werden.
public sealed record ParsedTrailerFilename(string Title, int Year, string ImdbId);

public static partial class TrailerFilename
{
    [GeneratedRegex(@"^trailer-(.+?)\s\((\d{4})\)\s\{imdb-id\s+(tt\d+)\}$")]
    private static partial Regex Pattern();

    /// <summary>Basename ohne Extension, z.B. "trailer-#9 (2009) {imdb-id tt0472033}".</summary>
    public static ParsedTrailerFilename? Parse(string basenameWithoutExt)
    {
        var match = Pattern().Match(basenameWithoutExt);
        if (!match.Success) return null;
        return new ParsedTrailerFilename(
            match.Groups[1].Value,
            int.Parse(match.Groups[2].Value),
            match.Groups[3].Value);
    }

    /// <summary>Baut aus den geparsten Teilen wieder exakt denselben
    /// Dateinamens-Stamm (ohne Extension).</summary>
    public static string BuildStem(ParsedTrailerFilename parsed) =>
        $"trailer-{parsed.Title} ({parsed.Year}) {{imdb-id {parsed.ImdbId}}}";
}
